src/ndb.c

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
+ * Contact: Roman Gezikov <rgezikov@exafore.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <string.h>
+#include "ndb.h"
+
+ndb_file_version_t ndb_file_version = 0x00000001;
+
+/* Period of repeating changes write operation, milliseconds */
+#define NDB_WRITE_PERIOD 500
+
+typedef Mutex ndb_mutex_t;
+typedef Thread *ndb_thread_t;
+
+static ndb_thread_t ndb_thread;
+static WORKING_AREA(ndb_thread_wa, 128);
+static msg_t nv_write_thread(void*);
+static ndb_mutex_t data_access;
+
+static enum ndb_op_code ndb_open_file(ndb_file_t *file, ndb_file_version_t *v);
+static enum ndb_op_code ndb_read(ndb_file_t *f, void *first_element, size_t s);
+
+#define OPTF __attribute__ ((weak, alias ("ndb_not_implemented")))
```

Comment:
FYI for the public APIs, there are a couple intricacies involved in getting this to work.
1.  `ndb_not_implemented()` needs to itself be *defined* as a weak symbol `__attribute__ ((weak))` in the header files where `__attribute__ ((weak, alias ("ndb_not_implemented")))` is applied to other function prototypes. I.e. you need something like 
```bool __attribute__ ((weak)) fake_unimplemented(void)
{
  return false;
}```
in the header file (probably a common header included by all the public API headers).
2. Said function prototypes with `__attribute__ ((weak, alias ("ndb_not_implemented")))` should NOT be visible to the source file where the actual implementations appear (otherwise you'll get redefinition errors).